### PR TITLE
[TASK] Add more type declarations

### LINKED ISSUE #564: Add more type declarations
We need to add parameter and return type declarations also in cases where the parent class or interface does not provide them. We need to add as many type declarations as possible without breaking compatibility with PHP 7.0, oelib 3.2.0, TYPO3 8LTS and TYPO3 9LTS.

This is required so we can add these type declarations in oelib 4.0.

### PR DESCRIPTION
**Pull request by Konafets**
Closes #564

### CODE REVIEW

**Comment by oliverklee** [comment]
Please rebase. I've just fixed the build failure related to the latest php-cs-fixer update.

File: Classes/BackEnd/AbstractEventMailForm.php [hunk 1]
```diff
@@ -201,7 +201,7 @@ private function validateFormData(): bool
      *
      * @return void
      */
-    public function markAsIncomplete()
+    public function markAsIncomplete(): void
```
Comment (oliverklee) [inline Classes/BackEnd/AbstractEventMailForm.php:204]: The `void` return type will need to wait until seminars 4.0 as this was introduced in PHP 7.1, but seminars 3 still needs to support PHP 7.0.

https://www.php.net/manual/de/migration71.new-features.php#migration71.new-features.void-functions

File: Classes/FrontEnd/DefaultController.php [hunk 2]
```diff
@@ -3482,7 +3482,7 @@ private function setVisibilityStatusMarker()
      *
      * @return void
      */
-    private function limitToTimeFrameSetting($builder)
+    private function limitToTimeFrameSetting(Tx_Seminars_BagBuilder_Event $builder)
```
Comment (oliverklee) [inline Classes/FrontEnd/DefaultController.php:3485]: To be safe for namespacing, please add a leading backslash to the calss name.

File: Classes/FrontEnd/DefaultController.php [hunk 3]
```diff
@@ -3615,7 +3615,7 @@ protected function redirectToCurrentUrl()
      */
     public function createSingleViewLink(
         \Tx_Seminars_Model_Event $event,
-        $linkText,
+        string $linkText,
         $htmlspecialcharLinkText = true
```
Comment (oliverklee) [inline Classes/FrontEnd/DefaultController.php:3619]: This one can have a `bool` type declaraton.

File: Classes/OldModel/AbstractModel.php [hunk 4]
```diff
@@ -309,7 +309,7 @@ protected function setRecordPropertyString(string $key, string $value)
      *
      * @return void
      */
-    protected function setRecordPropertyBoolean(string $key, $value)
+    protected function setRecordPropertyBoolean(string $key, bool $value)
```
Comment (oliverklee) [inline Classes/OldModel/AbstractModel.php:312]: The PHPDoc type annotation actually is incorrect here. The type annotation for `$value` needs to be `mixed`, and the type declaration then should not be there.

File: Classes/OldModel/Event.php [hunk 5]
```diff
@@ -2461,7 +2461,7 @@ public function isUserVip(int $userUid, int $defaultEventVipsFeGroupUid): bool
      *                 page, FALSE otherwise
      */
     public function canViewRegistrationsList(
-        $whichPlugin,
+        string $whichPlugin,
         $registrationsListPID = 0,
         $registrationsVipListPID = 0,
         $defaultEventVipsFeGroupID = 0,
```
Comment (oliverklee) [inline Classes/OldModel/Event.php:2467]: These other parameters also should have type declarations.

File: Classes/OldModel/Event.php [hunk 6]
```diff
@@ -2527,7 +2527,7 @@ public function canViewRegistrationsList(
      *                 page, FALSE otherwise
      */
     protected function canViewRegistrationsListForAttendeesAndManagersAccess(
-        $whichPlugin,
+        string $whichPlugin,
         $registrationsListPID = 0,
         $registrationsVipListPID = 0,
         $defaultEventVipsFeGroupID = 0
```
Comment (oliverklee) [inline Classes/OldModel/Event.php:2533]: Ditto.

File: Classes/OldModel/Event.php [hunk 7]
```diff
@@ -2602,7 +2602,7 @@ protected function canViewRegistrationsListForAttendeesAndManagersAccess(
      *                 page, FALSE otherwise
      */
     protected function canViewRegistrationsListForLoginAccess(
-        $whichPlugin,
+        string $whichPlugin,
         $registrationsListPID = 0,
         $registrationsVipListPID = 0,
         $defaultEventVipsFeGroupID = 0
```
Comment (oliverklee) [inline Classes/OldModel/Event.php:2608]: Ditto.

File: Classes/OldModel/Event.php [hunk 8]
```diff
@@ -2662,7 +2662,7 @@ protected function canViewRegistrationsListForLoginAccess(
      *                 page, FALSE otherwise
      */
     protected function canViewRegistrationsListForWorldAccess(
-        $whichPlugin,
+        string $whichPlugin,
         $registrationsListPID = 0,
         $registrationsVipListPID = 0,
         $defaultEventVipsFeGroupID = 0
```
Comment (oliverklee) [inline Classes/OldModel/Event.php:2668]: Ditto.

File: Classes/OldModel/Event.php [hunk 9]
```diff
@@ -2712,7 +2712,7 @@ protected function canViewRegistrationsListForWorldAccess(
      * @return string an empty string if everything is OK, a localized error
      *                error message otherwise
      */
-    public function canViewRegistrationsListMessage($whichPlugin, $accessLevel = 'attendees_and_managers'): string
+    public function canViewRegistrationsListMessage(string $whichPlugin, $accessLevel = 'attendees_and_managers'): string
```
Comment (oliverklee) [inline Classes/OldModel/Event.php:2715]: `$accesslevel` is a `string`.

File: Classes/SchedulerTasks/MailNotifier.php [hunk 10]
```diff
@@ -321,7 +321,7 @@ private function getCsv($eventUid): \Tx_Oelib_Attachment
      *
      * @return string the localized e-mail content, will not be empty
      */
-    private function customizeMessage($locallangKey, \Tx_Seminars_OldModel_Event $event, $organizerName = ''): string
+    private function customizeMessage(string $locallangKey, \Tx_Seminars_OldModel_Event $event, $organizerName = ''): string
```
Comment (oliverklee) [inline Classes/SchedulerTasks/MailNotifier.php:324]: `$organizerName` is a `string`.

File: Tests/LegacyUnit/FrontEnd/DefaultControllerTest.php [hunk 11]
```diff
@@ -6604,6 +6604,7 @@ public function registrationFormHtmlspecialcharsEventTitle()
         );
 
         $this->subject->piVars['seminar'] = $eventUid;
+        $this->subject->piVars['action'] = 'register';
```
Comment (oliverklee) [inline Tests/LegacyUnit/FrontEnd/DefaultControllerTest.php:6607]: Why the changes in this file?

**Review by oliverklee — CHANGES_REQUESTED** [review]

File: Tests/LegacyUnit/FrontEnd/DefaultControllerTest.php (hunk 11 above)
Comment (Konafets) [reply Tests/LegacyUnit/FrontEnd/DefaultControllerTest.php:6607]: Because otherwise [$registrationEditor->setAction($this->piVars['action']);](https://github.com/oliverklee/ext-seminars/blob/main/Classes/FrontEnd/DefaultController.php#L3016) will be empty and `setAction()` says in the header comment 
> * @param string $action action for which to create the form, must be either "register" or "unregister", must not be empty

File: Classes/OldModel/Event.php (hunk 9 above)
Comment (Konafets) [reply Classes/OldModel/Event.php:2715]: True, its a string but it passed as 2nd param to `canViewRegistrationsList($whichPlugin, $accessLevel)` but the 2nd param is an integer

```    
public function canViewRegistrationsList(
        string $whichPlugin,
        int $registrationsListPID = 0,
        int $registrationsVipListPID = 0,
        int $defaultEventVipsFeGroupID = 0,
        string $accessLevel = 'attendees_and_managers'
    ):
```

File: Classes/OldModel/AbstractModel.php (hunk 4 above)
Comment (oliverklee) [reply Classes/OldModel/AbstractModel.php]: Could you please update the PHPDoc annotation for `$value` to `mixed`?

File: Classes/OldModel/Event.php (hunk 9 above)
Comment (oliverklee) [reply Classes/OldModel/Event.php]: Thanks!

File: Tests/LegacyUnit/FrontEnd/DefaultControllerTest.php (hunk 11 above)
Comment (oliverklee) [reply Tests/LegacyUnit/FrontEnd/DefaultControllerTest.php:6607]: Then we need to fix the call to `setAction` instead, do a `??` check and a string cast around that.

Comment (Konafets) [reply Tests/LegacyUnit/FrontEnd/DefaultControllerTest.php:6607]: Like so 

```
        $action = $this->piVars['action'] ?? 'register';
        $registrationEditor->setAction((string) $action);
```

Comment (oliverklee) [reply Tests/LegacyUnit/FrontEnd/DefaultControllerTest.php]: Yes, or like this:
```php
$action = (string)($this->piVars['action'] ?? 'register');
$registrationEditor->setAction($action);
```
